landoapi/workers/revision_worker.py

```diff
@@ -0,0 +1,553 @@
+                    successor.status = RS.STALE
+    db.session.commit()
+
+    # Resolve dependency chain.
+    for revision in dependency_queue:
+        if revision.data["predecessor"]:
+            if len(revision.data["predecessor"]) == 1:
+                predecessor_revision = Revision.query.filter(
+                    Revision.revision_id == revision.data["predecessor"][0]
+                ).one()
+                revision.predecessor_id = predecessor_revision.id
+            if len(revision.data["predecessor"]) > 1:
+                revision.status = RS.PROBLEM
+                revision.update_data(error="Revision has more than one predecessor.")
+        else:
+            revision.predecessor = None
+    db.session.commit()
+
+
+def mark_stale_revisions():
+    """Discover any upstream changes, and mark revisions affected as stale."""
+    repos = Revision.query.with_entities(Revision.repo_name).distinct().all()
+    repos = tuple(repo[0] for repo in repos if repo[0])
+    for repo_name in repos:
+        repo = repo_clone_subsystem.repos[repo_name]
+        hgrepo = HgRepo(
+            str(repo_clone_subsystem.repo_paths[repo_name]),
+            config=repo.config_override,
+        )
+        # checkout repo, pull & update
+        with hgrepo.for_pull():
+            if hgrepo.incoming(repo.pull_path):
+                hgrepo.update_repo(repo.pull_path)
+                logger.info(f"Incoming changes detected in {repo_name}.")
+                revisions = Revision.query.filter(
+                    Revision.status.not_in(RS.LANDING_STATES),
+                    Revision.repo_name == repo_name,
+                )
+                logger.info(f"Marking {revisions.count()} revisions as stale.")
+                revisions.update({Revision.status: RS.STALE})
+    db.session.commit()
+
+
+class RevisionWorker(Worker):
+    """A worker that pre-processes revisions.
+
+    This worker continuously synchronises revisions with the remote Phabricator API
+    and runs all applicable checks and processes on each revision, if needed.
+    """
+
+    # DB configuration.
+    PAUSE_KEY = "REVISION_WORKER_PAUSED"
+    STOP_KEY = "REVISION_WORKER_STOPPED"
+    CAPACITY_KEY = "REVISION_WORKER_CAPACITY"
+    THROTTLE_KEY = "REVISION_WORKER_THROTTLE_SECONDS"
+
+    # Environment configuration.
+    IS_MAIN_KEY = "REVISION_WORKER_IS_MAIN"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(with_ssh=False, *args, **kwargs)
+
+    @property
+    def throttle_delay(self):
+        return ConfigurationVariable.get(self.THROTTLE_KEY, 1)
+
+    def throttle(self):
+        time.sleep(self.throttle_delay)
+
+    @property
+    def is_main(self):
+        """
+        Whether this worker is the main worker.
+
+        In case there are multiple workers running, this should ensure that only
+        one worker is discovering new revisions, and the rest are processing them.
+        """
+
+        return current_app.config[self.IS_MAIN_KEY]
```

Comment:
This should be instead use its own class, and in `cli` should have its own commands.

Workers can be called: `Supervisor` and `Processor`, both can be subclasses of `RevisionWorker`.